### PR TITLE
PaymentCustomizations Sample App (Stop Silencing Errors + Update)

### PR DESCRIPTION
**Pull request by mssalemi**
This PR updates the payment customization sample app in a few ways:
- The reorder and rename functions were removed (temporarily)
- The UI has been refactored to share more code between the create/edit pages
- User errors on save are now displayed at the top
- There is client-side validation now for the fields
- A delete button has been added to the edit page
- The function UI routes have been updated to use the function ID (i.e. `/hide/:functionId/:id` and `/hide/:functionId/new`)

Related issue: https://github.com/Shopify/script-service/issues/6008

### CODE REVIEW

File: sample-apps/payment-customization-sample-app/web/frontend/pages/index.jsx [hunk 1]
```diff
@@ -30,109 +29,97 @@ export default function HomePage() {
...
-    <EmptyState
-      heading="Hello world! 🎉"
-      {...props}
-    >
+    <EmptyState heading="Hello world! 🎉" {...props}>
       <p>
-        Welcome to the <b>Payment Customizations Functions Sample App</b>!
-        To get started, create a new customization.
+        Welcome to the <b>Payment Customizations Functions Sample App</b>! To
+        get started, create a new customization.
```
Comment (andrewhassan) [inline sample-apps/payment-customization-sample-app/web/frontend/pages/index.jsx:121]: Should we link to the payment customization settings here?

**Comment by dphm** [comment]
> `/<function-name>/:functionId/new`

I think this makes sense (PR description) but in practice the files are in `hide` 😆

The other 2 functions are rename and sort, so maybe you don't want the UI to all be `hide`-specific?

File: sample-apps/payment-customization-sample-app/web/frontend/hooks/useCustomizationForm.js [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
+  const initialData = props || {
+    cartSubtotal: 0,
+    paymentMethod: "Credit Card",
+  };
 
-  const [data, setData] = useState(initialData);
+  const [data, setData] = useState({
+    cartSubtotal: props?.cartSubtotal || initialData.cartSubtotal,
+    paymentMethod: props?.paymentMethod || initialData.paymentMethod,
+  });
```
[the remark this answers is not in the document]
Comment (mssalemi) [reply sample-apps/payment-customization-sample-app/web/frontend/hooks/useCustomizationForm.js:12]: Ty, makes sense and much cleaner.

File: sample-apps/payment-customization-sample-app/web/frontend/hooks/useCustomizationForm.js [hunk 3]
```diff
@@ -1,27 +1,29 @@
...
-  const {
-    initialData = {
-      cartSubtotal: 0,
-      paymentMethod: "",
-    },
-  } = props || {};
+  const initialData = {
+    cartSubtotal: 0,
+    paymentMethod: "Credit Card",
+    title: "HIDE",
```
Comment (dphm) [inline sample-apps/payment-customization-sample-app/web/frontend/hooks/useCustomizationForm.js:7]: All the new customizations are titled `HIDE` now but it used to be `Hide ${paymentMethod} if cart subtotal is bigger than ${cartSubtotal}`

Maybe `Hide ${paymentMethod} for orders over ${cartSubtotal}` would be a bit shorter?

File: sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationPageLayout.jsx [hunk 4]
```diff
@@ -22,9 +27,7 @@ export function CustomizationPageLayout({ actionProps, loading, children, ...pro
         breadcrumbs={[{ content: "Customizations", url: "/" }]}
```
Comment (dphm) [inline sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationPageLayout.jsx:27]: This page confused me a bit 😆 I got there from the back nav after creating a customization. It seems similar to the payment customizations index but slightly different, more like a "customization configurations" index? 🤔

The only action on that page is delete, but on this page you can activate/deactivate/create:

Screenshot 2022-12-15 at 11 09 33 AM

File: sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationForm.jsx [hunk 5]
```diff
@@ -42,7 +44,14 @@ export function CustomizationForm({
...
 
-        <Button submit primary loading={loading}>Save Customization</Button>
+        <Button
+          submit
+          primary
+          loading={loading}
+          disabled={!isNewCustomization && !hasChanged}
+        >
+          Save Customization
+        </Button>
```
Comment (dphm) [inline sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationForm.jsx:54]: I would prefer this button to be a [Page Action](https://polaris.shopify.com/components/page-actions)

**Review by dphm — COMMENTED** [review]
Not sure how much you're planning to fix more things in the app 😆 Left my comments anyway

File: sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationForm.jsx [hunk 6]
```diff
@@ -7,9 +8,36 @@ export function CustomizationForm({
...
+    setErrorBanner(null);
+    let validationErrors = [];
+
+    if (cartSubtotal < 0) {
+      validationErrors.push({
+        message: "Cart subtotal cannot be less than 0",
+      });
+    }
+
+    if (paymentMethod == "") {
```
Comment (andrewhassan) [inline sample-apps/payment-customization-sample-app/web/frontend/components/CustomizationForm.jsx:26]: Nit: To catch all whitespace, we can do `paymentMethod.trim() === ''`

File: sample-apps/payment-customization-sample-app/web/index.js [hunk 7]
```diff
@@ -144,6 +139,7 @@ export async function createServer(
 
       return { status: 200, data };
     } catch (err) {
+      console.log("error from here", err);
```
Comment (andrewhassan) [inline sample-apps/payment-customization-sample-app/web/index.js:142]: Don't forget to remove this!

File: sample-apps/payment-customization-sample-app/web/index.js [hunk 8]
```diff
@@ -265,40 +317,94 @@ export async function createServer(
...
 
-  app.get("/api/payment-customizations/:id", async (req, res) => {
+  app.put("/api/payment-customizations/:id", async (req, res) => {
+    const payload = req.body;
     const gid = idToGid(req.params.id);
 
-    const query = {
+    const { functionId, title, paymentMethod, cartSubtotal } = payload;
+
+    // update metafield
```
Comment (andrewhassan) [inline sample-apps/payment-customization-sample-app/web/index.js:334]: Nit: This comment isn't necessary

**Review by andrewhassan — APPROVED** [review]
🎩 and cleaned up some code

**Comment by DuncanUszkay1** [comment]
🎩 ed the function and UI, didn't read through the code

**Review by dphm — COMMENTED** [review]
Going to 🎩 now! Code looks nice!

**Review by dphm — APPROVED** [review]
🎩 Great!! The navigation makes more sense. Tested multiple user errors 👍🏽

File: sample-apps/payment-customization-sample-app/web/frontend/pages/hide/[functionId]/[id].jsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      };
+    }
+  };
+
+  if (isFetching || !data) return <Loading />;
+
+  return (
+    <CustomizationPage
+      title="Hide Payment Method"
+      subtitle="Hide a payment method for cart subtotal's greater than or equal to the specified subtotal."
```
Comment (dphm) [inline sample-apps/payment-customization-sample-app/web/frontend/pages/hide/[functionId]/[id].jsx:56]: Just saw this 😆 

```suggestion
      subtitle="Hide a payment method for cart subtotals greater than or equal to the specified subtotal."
```

File: sample-apps/payment-customization-sample-app/web/frontend/pages/hide/[functionId]/new.jsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+          { message: "An unexpected error occurred. Please try again later." },
+        ],
+      };
+    }
+  };
+
+  return (
+    <CustomizationPage
+      title="Hide Payment Method"
+      subtitle="Hide a payment method for cart subtotal's greater than or equal to the specified subtotal."
```
Comment (dphm) [inline sample-apps/payment-customization-sample-app/web/frontend/pages/hide/[functionId]/new.jsx:50]:
```suggestion
      subtitle="Hide a payment method for cart subtotals greater than or equal to the specified subtotal."
```

**Pull request merged**